pkg/apis/kubeflow/v1alpha2/common_types.go

```diff
@@ -0,0 +1,156 @@
+	// will be created for this replica. RestartPolicy in PodTemplateSpec
+	// will be overide by RestartPolicy in ReplicaSpec
+	Template v1.PodTemplateSpec `json:"template,omitempty"`
+
+	// Restart policy for all replicas within the job.
+	// One of Always, OnFailure, Never and ExitCode.
+	// Default to Never.
+	RestartPolicy RestartPolicy `json:"restartPolicy,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+// JobCondition describes the state of the job at a certain point.
+type JobCondition struct {
+	// Type of job condition.
+	Type JobConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+	// A human readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+// JobConditionType defines all kinds of types of JobStatus.
+type JobConditionType string
+
+const (
+	// JobCreated means the job has been accepted by the system,
+	// but one or more of the pods/services has not been started.
+	// This includes time before pods being scheduled and launched.
+	JobCreated JobConditionType = "Created"
+
+	// JobRunning means all sub-resources (e.g. services/pods) of this job
+	// have been successfully scheduled and launched.
+	// The training is running without error.
+	JobRunning JobConditionType = "Running"
+
+	// JobRestarting means one or more sub-resources (e.g. services/pods) of this job
+	// reached phase failed but maybe restarted according to it's restart policy
+	// which specified by user in v1.PodTemplateSpec.
+	// The training is freezing/pending.
+	JobRestarting JobConditionType = "Restarting"
+
+	// JobSucceeded means all sub-resources (e.g. services/pods) of this job
+	// reached phase have terminated in success.
+	// The training is complete without error.
+	JobSucceeded JobConditionType = "Succeeded"
+
+	// JobFailed means one or more sub-resources (e.g. services/pods) of this job
+	// reached phase failed with no restarting.
+	// The training has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+
+// CleanPodPolicy describes how to deal with pods when the job is finished.
+type CleanPodPolicy string
+
+const (
+	CleanPodPolicyUndefined CleanPodPolicy = ""
+	CleanPodPolicyAll       CleanPodPolicy = "All"
+	CleanPodPolicyRunning   CleanPodPolicy = "Running"
+	CleanPodPolicyNone      CleanPodPolicy = "None"
+)
+
+// RestartPolicy describes how the replicas should be restarted.
+// Only one of the following restart policies may be specified.
+// If none of the following policies is specified, the default one
+// is RestartPolicyAlways.
```

Comment:
Let's focus on MPIJob's spec for this PR to avoid any divergent changes in common types. I'll keep these notes unresolved so we can address them later.